### PR TITLE
feat: implement checkpoint download and resume from S3

### PR DESCRIPTION
**Pull request by efazal**
**What this PR does / why we need it**:

Adds S3 object storage support for TransformersTrainer checkpoints with automatic download before training and resume functionality. 

  - S3 checkpoint download: Automatically downloads latest checkpoint from S3 before training starts (rank-0-only with barrier synchronization)
  - Incomplete checkpoint detection: Skips incomplete checkpoints marked with sentinel files during download
  - Auto-resume from S3: Integrates with existing auto-resume logic to restore from S3-stored checkpoints
  - Cloud storage validation: Validates S3 bucket accessibility and credentials during initialization

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

## Summary by CodeRabbit

* **New Features**
  * S3‑compatible remote checkpoint support: optional cloud URI, rank‑0 download of latest valid checkpoint with progress reporting, barrier‑based distributed synchronization, and storage-access verification (test write/read).
  * Output directory normalization and validation for cloud/local URIs.
  * New flags to enable/disable cloud storage access checks and SSL verification; warns when SSL verification is disabled.

* **Tests**
  * Integration tests for S3 checkpoint download: latest‑checkpoint selection, incomplete checkpoint handling, and distributed scenarios.

* **Chores**
  * Optional runtime support for fsspec/s3fs added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by efazal** [comment]
@robert-bell @hrathina please review.

File: kubeflow/trainer/rhai/transformers.py [hunk 1]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
     class JITCheckpointCallback(TrainerCallback):
         """Transformers callback that integrates JIT checkpointing with trainer lifecycle."""
 
-        def __init__(self):
+        def __init__(self, cloud_remote_storage_uri=None):
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: type annotations?

File: kubeflow/trainer/rhai/transformers.py [hunk 2]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+                bucket = self.cloud_storage_base_path.split("/")[0]
+
+                fsspec_kwargs = {}
+                if self.cloud_storage_protocol == "s3":
+                    # AWS_S3_ENDPOINT must be explicitly passed since it's not
+                    # a standardized name like other AWS credentials
+                    endpoint_url = os.environ.get("AWS_S3_ENDPOINT")
+                    if endpoint_url:
+                        fsspec_kwargs = {
+                            "client_kwargs": {"endpoint_url": endpoint_url, "verify": False},
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: question: why `verify: false`?

File: kubeflow/trainer/rhai/transformers.py [hunk 3]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+            self.fsspec = None
+            self.cloud_storage_protocol = None
+            self.cloud_storage_base_path = None
+
+            if cloud_remote_storage_uri and "://" in cloud_remote_storage_uri:
+                import fsspec
+
+                self.cloud_storage_protocol = cloud_remote_storage_uri.split("://")[0]
+                self.cloud_storage_base_path = cloud_remote_storage_uri.split("://", 1)[1]
+                bucket = self.cloud_storage_base_path.split("/")[0]
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: issue: can you check what happens if the user adds a slash to the end of the checkpoint storage? There's usually some edge cases that need checking. It'd be worth adding some unit tests for it. Here's the usual odd candidates... but there might be some more.

E.g.s
```python
output_dirs = [
  "s3://my-bucket",
  "s3://my-bucket/",  
  "s3://my-bucket//",
  "s3://my-bucket/prefix",
  "s3://my-bucket/prefix/",
  "s3://my-bucket/prefix//",
]
```

File: kubeflow/trainer/rhai/transformers.py [hunk 4]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+                    endpoint_url = os.environ.get("AWS_S3_ENDPOINT")
+                    if endpoint_url:
+                        fsspec_kwargs = {
+                            "client_kwargs": {"endpoint_url": endpoint_url, "verify": False},
+                            "config_kwargs": {"signature_version": "s3v4"},
+                        }
+
+                try:
+                    # Verify connection and confirm bucket exists
+                    self.fsspec = fsspec.filesystem(self.cloud_storage_protocol, **fsspec_kwargs)
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: suggestion(optional): I think fsspec has a way to embed the prefix in the fs, which might let you remove `cloud_storage_base_path`

```python
fs = fsspec.filesystem("dir", path="my-bucket/prefix", fs=fsspec.filsystem("s3"))
```

then to get object at `s3://my-bucket/prefix/foo/bar` you do something like `fs.open("foo/bar")`

File: kubeflow/trainer/rhai/transformers.py [hunk 5]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+                    self.fsspec.info(f"{self.cloud_storage_protocol}://{bucket}")
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: issue: what happens if the user only has permission to read a subpath in the bucket, but not the bucket itself?

How about writing a small file and checking we can read it? (But don't delete it, as we don't need to check that permission.)

We probably ought to have a flag in the TransformersTrainer for disabling this check too, just in case it's a false positive. It should default to on, but explain in the error message how to disable it, e.g. `this check can be disabled by setting ....`.

File: kubeflow/trainer/rhai/transformers.py [hunk 6]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
     class JITCheckpointCallback(TrainerCallback):
         """Transformers callback that integrates JIT checkpointing with trainer lifecycle."""
 
-        def __init__(self):
+        def __init__(self, cloud_remote_storage_uri=None):
             self.jit_manager = None
             self._trainer_ref = None
+            self.cloud_remote_storage_uri = cloud_remote_storage_uri
+            self.fsspec = None
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: can we find a better name than `fsspec`? how about `remote_fs`?

File: kubeflow/trainer/rhai/transformers.py [hunk 7]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+            """Download latest checkpoint from S3 before training (rank-0-only with barrier)."""
+            if not self.fsspec:
+                return
+
+            is_rank_0 = state.is_world_process_zero
+
+            if is_rank_0:
+                checkpoint_paths = self.fsspec.glob(f"{self.cloud_storage_base_path}/checkpoint-*")
+                steps = sorted(
+                    {int(p.split("checkpoint-")[1].split("/")[0]) for p in checkpoint_paths},
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: `int(p.split("checkpoint-")[1].split("/")[0])` might throw.

Maybe use a regex instead? e.g. `r"^foo-(\d+)$"`

File: kubeflow/trainer/rhai/transformers.py [hunk 8]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+                for step in steps:
+                    name = f"checkpoint-{step}"
+                    marker = f"{self.cloud_storage_base_path}/{name}/{CHECKPOINT_INCOMPLETE_MARKER}"
+                    if self.fsspec.exists(marker):
+                        continue
+
+                    print(f"[Kubeflow] Downloading checkpoint: {name}", flush=True)
+                    self.fsspec.get(
+                        f"{self.cloud_storage_base_path}/{name}", args.output_dir, recursive=True
+                    )
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: suggestion: can we print out the files it's downloading so users can see the progress? This'll be especially important if there's a big checkpoint or the network speed is very slow.

File: kubeflow/trainer/rhai/transformers.py [hunk 9]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+                    raise RuntimeError(
+                        f"Permission denied accessing bucket '{bucket}' "
+                        f"in {self.cloud_storage_protocol} storage. "
+                        f"Please verify your credentials are configured correctly."
+                    ) from e
+
+        def on_init_end(self, args, state, control, **kwargs):
+            """Download latest checkpoint from S3 before training (rank-0-only with barrier)."""
+            if not self.fsspec:
+                return
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: worth adding a comment saying why we're just returning here?

File: kubeflow/trainer/rhai/transformers.py [hunk 10]
```diff
@@ -323,9 +324,88 @@ def checkpoint_in_progress(self):
...
+                    break
+
+            # Barrier to wait for rank 0 checkpoint download to complete
+            try:
+                if dist.is_initialized():
+                    dist.barrier()
+            except Exception as e:
+                print(
+                    f"[Kubeflow] Warning: Barrier failed after checkpoint download: {e}", flush=True
+                )
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: question: what happens if this fails? Are we able to just carry on?

**Review by robert-bell — APPROVED** [review]
thanks @efazal! lgtm. A few minor suggestions an edge cases to check, but looks good.

@hrathina please do take a look too, particularly at the tests which I've only briefly glanced at.

File: kubeflow/trainer/rhai/transformers.py (hunk 2 above)
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: This is for s3 compatible storages like Minio and similar where they are typically self deployed within some cloud tenant and use self signed certs. Perhaps, we should set it to True. And make it configurable via env for the user to disable it. Happened to me during my testing with Minio, where it's deployed in AWS.

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py]: Ah understood.

>  Perhaps, we should set it to True

My preference is we leave it at the default, but provide users a way to override the settings passed -- i.e. let users pass an extra dict of settings to the TransformersTrainer that gets passed here.

```trainer = TransformersTrainer(
    func=...,
    output_dir="s3://my-bucket/prefix",
    cloud_storate_settings={
       "validate": False,
       "endpoint_url": "https://minio.minio-ns.svc" 
    },
)
```

What does the data connection secret provide us? Does it contain a setting for `validate`? Does it contain the `endpoint_url`?

Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Yeah I agree. So the default is validate true. The data connection provides secret, key, region, bucket name, and enpoint_url only. Doesn't provide validate.

Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py]: OK - I reckon we should remove that flag for this PR.

For supporting the extra config, I'm happy to pull that out into a separate task. I think it's something that needs doing for GA - wdyt?

File: kubeflow/trainer/rhai/transformers.py [hunk 11]
```diff
@@ -92,6 +92,11 @@ class TransformersTrainer:
                               Data Science project, go to the Connections tab, and either
                               copy an existing connection's resource name or create a new
                               S3-compatible connection.
+        verify_storage_access: Verify storage access by writing/reading a test file.
+                              Default: True. Can be disabled if you get false positives.
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: can we improve this docstring? It's not clear what the storage is, precisely what it does, and why it should be kept enabled. Changing the parameter name might help too.

File: kubeflow/trainer/rhai/transformers.py [hunk 12]
```diff
@@ -92,6 +92,11 @@ class TransformersTrainer:
...
+        verify_ssl: Verify SSL certificates for S3 connections. Default: True.
+                   Set to False only if using S3-compatible storage with self-signed
+                   certificates. WARNING: Disabling SSL verification is a security risk.
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: this needs a better parameter name. It's only verifying ssl for the cloud storage for checkpoints.

File: kubeflow/trainer/rhai/transformers.py [hunk 13]
```diff
@@ -323,9 +332,143 @@ def checkpoint_in_progress(self):
...
+                fsspec_kwargs = {}
+                if protocol == "s3":
+                    # AWS_S3_ENDPOINT must be explicitly passed since it's not
+                    # a standardized name like other AWS credentials
+                    endpoint_url = os.environ.get("AWS_S3_ENDPOINT")
+                    if endpoint_url:
+                        verify_ssl = checkpoint_config.get("verify_ssl", True)
+                        fsspec_kwargs = {
+                            "client_kwargs": {"endpoint_url": endpoint_url, "verify": verify_ssl},
+                            "config_kwargs": {"signature_version": "s3v4"},
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: question: will users ever need to change signature_version?

File: kubeflow/trainer/rhai/transformers.py [hunk 14]
```diff
@@ -323,9 +332,143 @@ def checkpoint_in_progress(self):
...
             self.jit_manager = None
             self._trainer_ref = None
+            self.cloud_remote_storage_uri = cloud_remote_storage_uri
+            self.remote_fs = None
+
+            if cloud_remote_storage_uri and "://" in cloud_remote_storage_uri:
+                import fsspec
+
+                protocol = cloud_remote_storage_uri.split("://")[0]
+                base_path = cloud_remote_storage_uri.split("://", 1)[1]
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: ?

```suggestion
                protocol, base_path = cloud_remote_storage_uri.split("://", 1)
```

**Review by hrathina — APPROVED** [review]

File: kubeflow/trainer/rhai/transformers.py [hunk 15]
```diff
@@ -323,9 +332,143 @@ def checkpoint_in_progress(self):
...
+                        f"[Kubeflow] Cloud storage configured: {cloud_remote_storage_uri} "
+                        f"({protocol})",
+                        flush=True,
+                    )
+                except Exception as e:
+                    raise RuntimeError(
+                        f"Cannot access storage path: {cloud_remote_storage_uri}. "
+                        f"Error: {e}. "
+                        f"If using self-signed certificates, set verify_ssl=False. "
+                        f"If experiencing permission issues, set verify_storage_access=False."
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: suggestion:

```suggestion
                        f"Failed to check this node has access to the storage path: '{cloud_remote_storage_uri}'. "
                        f"Error: {e}. "
                        f"If using self-signed certificates, set verify_ssl=False. "
                        f"If experiencing permission issues, check you have read and write permissions to '{cloud_remote_storage_uri}'. "
                        f"This check can be disabled by setting verify_storage_access=False."
```

File: kubeflow/trainer/rhai/transformers.py [hunk 16]
```diff
@@ -323,9 +332,143 @@ def checkpoint_in_progress(self):
...
+                    print("[Kubeflow] Download complete", flush=True)
+                    break
+
+            # Barrier to wait for rank 0 checkpoint download to complete
+            try:
+                if dist.is_initialized():
+                    dist.barrier()
+            except Exception as e:
+                raise RuntimeError(
+                    "[Kubeflow] Barrier synchronization failed during checkpoint download."
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: question: what does this mean for users? What should they do if they see this error?

File: kubeflow/trainer/rhai/transformers.py (hunk 2 above)
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: I've already added two new fields in the TrainerformerTrainer dataclass, to `verify_ssl` and `verify_stroage_access`, that get's wired into the instrumentation code. With messages to the let the user know either can be disabled if needed. Is that okay?

**Review by robert-bell — APPROVED** [review]
Thanks @efazal the updates look great. I particularly like the improved error and log messages, and the progress callback on the download. It all really helps make for a great user experience.

I've added a few more suggestions, but again I'm happy for it to be merged once @hrathina has reviewed.

File: kubeflow/trainer/rhai/transformers.py [hunk 17]
```diff
@@ -121,6 +126,8 @@ class TransformersTrainer:
     output_dir: Optional[str] = None
     periodic_checkpoint_config: Optional[PeriodicCheckpointConfig] = None
     data_connection_name: Optional[str] = None
+    verify_storage_access: bool = True
```
Comment (hrathina) [inline kubeflow/trainer/rhai/transformers.py]: nit: it is good if we haev verify_cloud_storage_access instead of verify_storage_access

File: kubeflow/trainer/rhai/transformers.py (hunk 13 above)
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Most likely not, only extremely old s3 compatible deployments pre 2026, would need to configure, otherwise no. In fact, the default is anyway s3v4. I'll remove the signature_version.

File: kubeflow/trainer/rhai/transformers.py (hunk 16 above)
Comment (efazal) [reply kubeflow/trainer/rhai/transformers.py]: Most likely a rank synchronization related issue or related related to training configuration. I believe the torch distributed will throw an appropriate error prompting the user to fix something related to the training. Oh, I'm not printing the exception. I shall fix it with a better message as well

**Comment by robert-bell** [comment]
@efazal I forgot to add -- should we add `fsspec` and `s3fs` to the project requirements? I think probably ought to.

I'm wondering whether we should add an extra `optional-dependencies` group called `rhai`? We'll need to update the runtime images to install this dependency.

File: pyproject.toml [hunk 18]
```diff
@@ -39,6 +39,10 @@ docker = [
 podman = [
   "podman>=5.6.0"
 ]
+rhai = [
+  "fsspec>=2025.9.0",
+  "s3fs>=2024.12.0",
```
Comment (robert-bell) [inline pyproject.toml:44]: @briangallagher @kramaranya @Fiona-Waters @MStokluska could we get your review on this please? It's a change outside the directory the KFT team owns.

To save checkpoints to s3, we need these extra deps. I think they're already installed in the runtime images so this shouldn't cause any conflicts.

@efazal can you create a ticket to remind us to update the runtime images so we're installing `kubeflow-sdk[rhai]`?

Comment (Fiona-Waters) [reply pyproject.toml:44]: This looks okay to me. Any thoughts @kramaranya @briangallagher @MStokluska ?

**Review by briangallagher — APPROVED** [review]
/lgtm
/approve
